File: django_archimage/static/scope/js/archimage/archimage.js

// Use this to hit the api:
// "command" is required, arguments is optional
// args is a "stringify-ed" json object
function SendCommand(command,arguments) {
  var jqXHR = $.ajax({
   type:"POST",
   url:"/api/execute/",
   data: {
    'command': command,
    'args': arguments,
    'csrfmiddlewaretoken': document.getElementsByName('csrfmiddlewaretoken')[0].value
  },
  dataType: 'json'
});
};

// mouse clicked?
var is_clicked = false
$(document).mousedown(function(e) {
  is_clicked = true
});
$(document).mouseup(function(e) {
  is_clicked = false
});

// -----------------------------------------------------------------------------------------------------
//                                     Misc Mount Controls
// -----------------------------------------------------------------------------------------------------
// show-hide shortcuts
$(function(){
  $('#toggle-shortcuts').change(function() { // TOGGLE: Connection
    if ($(this).prop('checked')) { // show
      $('#collapseExample').collapse('show')
    } else { // hide
      $('#collapseExample').collapse('hide')
    }
  });
})
// Serial connection toggle
$(function(){
    $('#toggle-connection').change(function() { // TOGGLE: Connection
        if ($(this).prop('checked')) { // connect
          SendCommand('connect',null);
            } else { // disconnect
              SendCommand('disconnect',null);
            }
          });
  })

// Helper functions:
$(function(){
  $("#align_level_south").mousedown(function(){
        SendCommand('init_align',null); // Send the command to start moving
      });
  $("#clear_tracking_rates").mousedown(function(){
        SendCommand('clear_track_rates',null); // Send the command to start moving
      });
});

// Home:
$(function(){
  $("#home_setup").mousedown(function(){
        SendCommand('home_setup',null); // Send the command to start moving
      });
  $("#home_find").mousedown(function(){
        SendCommand('find_home',null); // Send the command to start moving
      });
});

// Park:
$(function(){
  $("#park_set").mousedown(function(){
        SendCommand('set_park',null); // Send the command to start moving
      });
  $("#park").mousedown(function(){
        SendCommand('home_park',null); // Send the command to start moving
      });
});

// -----------------------------------------------------------------------------------------------------
//                                     Polling Example
// -----------------------------------------------------------------------------------------------------
// Mount Monitor Poll
(function poll() {
  setTimeout( function() {
    if ($('#toggle-monitor').prop('checked')){
      $('#monitor-display').show();
      $.ajax({
       type:"POST",
       url:"/api/execute/",
       data: {
        'command': 'get_all',
        'args': null,
        'csrfmiddlewaretoken': document.getElementsByName('csrfmiddlewaretoken')[0].value
      },
      dataType: 'json',
      success: function (result) {
              // update the display
              $('#pointing_readout').text(JSON.stringify(result.response.pointing, null, 4));
              $('#tracking_readout').text(JSON.stringify(result.response.tracking, null, 4));
              $('#target_readout').text(JSON.stringify(result.response.target, null, 4));
              $('#status_readout').text(JSON.stringify(result.response.status, null, 4));
              // poll again
              poll();
            }
          });
    } else {
      // poll again
      //$('#pointing_readout').text('Monitoring Disabled');
      $('#monitor-display').hide();
      poll();
    }
  }, 1000);
})();

// -----------------------------------------------------------------------------------------------------
//                                      Paddle Controls
// -----------------------------------------------------------------------------------------------------
// Paddle controls:
// - commands delivered via data-downcommand and data-upcommand (press and release),
// - id='paddle_something' to match contains jsquery for 'paddle'
$(function(){
  $("[id*='paddle']").bind('touchstart mousedown', function() {
        SendCommand($(this).data("downcommand"),null); // Send the command to start moving
      });
  $("[id*='paddle']").bind('touchend mouseup', function() {
        SendCommand($(this).data("upcommand"),null); // Send the command to start moving
      });
  $("[id*='stop']").bind('touchstart mousedown', function() {
        SendCommand($(this).data("downcommand"),null); // Send the command to start moving
      });
});
// jog speed slider
$(function(){
  var paddleslider = $('#paddle-speed-slider').bootstrapSlider({
    tooltip: 'always',
    tooltip_position: 'bottom',
    precision: 2,
    data: 'slider',
    formatter: function(value) {
      //$('#paddle-speed-header').text('Speed: ' + value);
      return value;
    }
  });
  $("#paddle-speed-slider").bind('touchend mouseup', function() {
        //slide = $("#paddle-speed-slider").bootstrapSlider();
        speed = paddleslider.bootstrapSlider('getValue');
        SendCommand('set_speed',JSON.stringify({slew_speed_ds:speed}));
        $('#paddle-speed-header').text('Speed: ' + speed + ' deg/s');
      });
});

// -----------------------------------------------------------------------------------------------------
//                              Location Input Shtuff
// -----------------------------------------------------------------------------------------------------
// Location Stuff
$(function(){

  $("#location-btn").bind('touchstart mousedown', function() {
      // Set latitude
      var lat = parseFloat($("#latitude-input").val())
      var lon = parseFloat($("#longitude-input").val())
      console.log('Setting latitude to ' + lat)
      console.log('Setting longitude to ' + lon)
      SendCommand('set_location',JSON.stringify({lat_dd:lat,lon_dd:lon}))

    });
});

// -----------------------------------------------------------------------------------------------------
//                              GOTO Target Specification Shtuff
// -----------------------------------------------------------------------------------------------------
$(function(){
  change_goto_coordinates()
  check_goto_coordinates()
  $("#goto-target-btn").bind('touchstart mousedown',submit_goto_target);
  $('#toggle-goto-angle-format').change(update_goto);
  $('#goto-coordinate-id').change(update_goto);
});

function submit_goto_target(){
  if($('#goto-coordinate-id').val()=="ra-dec"){ //celestial pointing
    if ($("#toggle-goto-angle-format").prop('checked')) { // degrees
      var ra_dd = parseFloat($("#goto-ra-degrees").val());
      var dec_dd = parseFloat($("#goto-dec-degrees").val());
    } else { // sexasigmal
      var ra_h = parseFloat($("#goto-ra-hours").val());
      var ra_m = parseFloat($("#goto-ra-minutes").val());
      var ra_s = parseFloat($("#goto-ra-seconds").val());
      var ra_dd = ra_h*15 + ra_m/4 + ra_s/240;
      var dec_d = parseFloat($("#goto-dec-degrees").val());
      var dec_m = parseFloat($("#goto-dec-minutes").val());
      var dec_s = parseFloat($("#goto-dec-seconds").val());
      var dec_dd = dec_d + dec_m/60 + dec_s/3600
    }
    SendCommand('point_radec',JSON.stringify({ra_deg:ra_dd,dec_deg:dec_dd}))
  } else { //topocentric pointing
    if ($("#toggle-goto-angle-format").prop('checked')) { // degrees
      var alt_dd = parseFloat($("#goto-alt-degrees").val());
      var az_dd = parseFloat($("#goto-az-degrees").val());
    } else { // sexasigmal
      var alt_d = parseFloat($("#goto-alt-degrees").val());
      var alt_m = parseFloat($("#goto-alt-minutes").val());
      var alt_s = parseFloat($("#goto-alt-seconds").val());
      var alt_dd = alt_d + alt_m/60 + alt_s/3600
      var az_d = parseFloat($("#goto-az-degrees").val());
      var az_m = parseFloat($("#goto-az-minutes").val());
      var az_s = parseFloat($("#goto-az-seconds").val());
      var az_dd = az_d + az_m/60 + az_s/3600
    }
    SendCommand('point_altaz',JSON.stringify({alt_deg:alt_dd,az_deg:az_dd}))
  }
  return
}

// wrapper for all the display updating
function update_goto(){
  change_goto_coordinates()
  check_goto_coordinates()
  return
}
// display inputs for the selected coordniate SYSTEM
function change_goto_coordinates(){
  if($('#goto-coordinate-id').val()=="ra-dec"){
    $('#goto-topo').hide()
    $('#goto-celestial').show()
  } else {
    $('#goto-celestial').hide()
    $('#goto-topo').show()
  }
}
// display inputs for the selected coordniate FORMAT
function check_goto_coordinates(){
  if($('#goto-coordinate-id').val()=="ra-dec"){ //celestial pointing
    if ($("#toggle-goto-angle-format").prop('checked')) { // degrees
      // hour angle
      $("#goto-ra-degrees-form").show();
      $("#goto-ra-minutes-form").hide();
      $("#goto-ra-seconds-form").hide();
      $("#goto-ra-hours-form").hide();
      // declination
      $("#goto-dec-degrees-form").show();
      $("#goto-dec-minutes-form").hide();
      $("#goto-dec-seconds-form").hide();
    } else { // sexasigmal
      // hour angle
      $("#goto-ra-degrees-form").hide();
      $("#goto-ra-hours-form").show();
      $("#goto-ra-minutes-form").show();
      $("#goto-ra-seconds-form").show();
      // declination
      $("#goto-dec-degrees-form").show();
      $("#goto-dec-minutes-form").show();
      $("#goto-dec-seconds-form").show();
    }
  } else { //topocentric pointing
    if ($("#toggle-goto-angle-format").prop('checked')) { // degrees
      // azimiuth
      $("#goto-az-degrees-form").show();
      $("#goto-az-minutes-form").hide();
      $("#goto-az-seconds-form").hide();
      // altitude
      $("#goto-alt-degrees-form").show();
      $("#goto-alt-minutes-form").hide();
      $("#goto-alt-seconds-form").hide();
    } else { // sexasigmal
      // azimiuth
      $("#goto-az-degrees-form").show();
      $("#goto-az-minutes-form").show();
      $("#goto-az-seconds-form").show();
      // altitude
      $("#goto-alt-degrees-form").show();
      $("#goto-alt-minutes-form").show();
      $("#goto-alt-seconds-form").show();
    }
  }
  return
}

// -----------------------------------------------------------------------------------------------------
//                              Coordinate Specification Shtuff
// -----------------------------------------------------------------------------------------------------
$(function(){
  $("#pointing-btn").bind('touchstart mousedown',submit_pointing);
  $('#toggle-angle-format').change(check_coordinates);
  $('#coordinate-id').change(check_coordinates);
});

function submit_pointing(){
  var degrees = 0.0
  var hours = 0.0
  var minutes = 0.0
  var seconds = 0.0
  var coord = $('#coordinate-id').val();

  if ($("#toggle-angle-format").prop('checked')) { // degrees
    degrees = parseFloat($("#degrees").val())
  } else { // sexasigmal
    minutes = parseFloat($("#minutes").val())
    seconds = parseFloat($("#seconds").val())
    if($('#coordinate-id option:selected').text()=='Hour Angle'){ // hours
      hours = parseFloat($("#hours").val())
      degrees = hours*15 + minutes/4 + seconds/240
    } else { //degrees
      d = parseFloat($("#degrees").val())
      degrees  = d + minutes/60 + seconds/3600
    }
  }
  key = coord + '_dd'
  SendCommand('set_pointing_'+coord,JSON.stringify({[key]:degrees}))
  return
}

function check_coordinates(){
  if ($("#toggle-angle-format").prop('checked')) { // degrees
    $("#minutes-form").hide();
    $("#seconds-form").hide();
    $("#hours-form").hide();
    $("#degrees-form").show();
  } else { // sexasigmal
    if($('#coordinate-id').val()=='ha'){
      $("#degrees-form").hide();
      $("#hours-form").show();
    } else {
      $("#hours-form").hide();
      $("#degrees-form").show();
    }
    $("#minutes-form").show();
    $("#seconds-form").show();
  }
  return
}

// -----------------------------------------------------------------------------------------------------
//                                     Simulator Controls
// -----------------------------------------------------------------------------------------------------
$(function(){
  var sim_ra_position_slider = $('#simulator-ra-position-slider').bootstrapSlider({
    tooltip: 'always',
    tooltip_position: 'bottom',
    precision: 2,
    data: 'slider',
    formatter: function(value) {
      //$('#paddle-speed-header').text('Speed: ' + value);
      return value;
    }
  });
  $("#simulator-ra-position-slider").bind('touchmove mousemove touchend mouseup', function() {
    if (is_clicked) {
      ra = sim_ra_position_slider.bootstrapSlider('getValue')
      SendCommand('set_virtual_ra',JSON.stringify({ra_dd:ra}));
    }
  });
  var sim_dec_position_slider = $('#simulator-dec-position-slider').bootstrapSlider({
    tooltip: 'always',
    tooltip_position: 'bottom',
    precision: 2,
    data: 'slider',
    formatter: function(value) {
      return value;
    }
  });
  $("#simulator-dec-position-slider").bind('touchmove mousemove touchend mouseup', function() {
    if (is_clicked) {
      dec = sim_dec_position_slider.bootstrapSlider('getValue');
      SendCommand('set_virtual_dec',JSON.stringify({dec_dd:dec}));
    }

  });
});


// -----------------------------------------------------------------------------------------------------
//                                     Moount Flag Controls
// -----------------------------------------------------------------------------------------------------
// Archiamge flag toggles
$(function() {

  $('[id^=toggle-tracking]').change(function() { // TOGGLE: Track
    if ($(this).prop('checked')) { // activate track
      $(this).bootstrapToggle('On');
      SendCommand('enable_track',null);
    } else { // disable track
      SendCommand('disable_track',null);
    }
  });

  $('#toggle-configuration').change(function() { // TOGGLE: Mount Configuration
    if ($(this).prop('checked')) { // it's equatorial
      SendCommand('set_equatorial',null);
      $('#toggle-german').bootstrapToggle('enable'); // Enable german equatorial toggle if alt/az
    } else { // else it's alt/az
      $('#toggle-german').prop('checked',false).change(); // Turn off german equatorial if alt/az
      $('#toggle-german').bootstrapToggle('disable'); // Disable german equatorial toggle if alt/az
      SendCommand('set_altaz',null);
      SendCommand('disable_german_equatorial',null)
    }
  });

  $('#toggle-ieca').change(function() { // TOGGLE: IECA
    if ($(this).prop('checked')) { // activate IECA
      SendCommand('enable_ieca',null);
    } else { // disable IECA
      SendCommand('disable_ieca',null);
    }
  });

  $('#toggle-pec').change(function() { // TOGGLE: PEC
    if ($(this).prop('checked')) { // activate PEC
      SendCommand('enable_pec',null);
    } else { // disable PEC
      SendCommand('disable_pec',null);
    }
  });

  $('#toggle-horrizon').change(function() { // TOGGLE: Horrizon Limit
    if ($(this).prop('checked')) { // activate Horrizon Limit
      SendCommand('enable_horrizon_limit',null);
    } else { // disable Horrizon Limit
      SendCommand('disable_horrizon_limit',null);
    }
  });

  $('#toggle-german').change(function() { // TOGGLE: German Equatorial
    if ($(this).prop('checked')) { // activate german equatorial
      SendCommand('enable_german_equatorial',null);
    } else { // deactivate german equatorial
      SendCommand('disable_german_equatorial',null);
    }
  });

})

File: archimage.py
#!python2
import serial
import ephem
import io, re, random, time, binascii, json
from datetime import datetime

class archimage():
    def __init__(self,live=True):
        '''
        I added critical stuff to the init.  Things like comms, basic  settings required for propper alignment.... etc.
        '''
        self.debug_flag = False
        
        # location:
        self.lat = 39.94963
        self.lon = -83.81756
        self.alt = 316
        
        # astronomy stuff:
        self.sidereal_rate = 15.041 # arcseconds/second

        # Serial Comm Stuff
        self.live_comm = live # use for debugging without comms hooked up
        self.port = 'COM3'
        self.baudrate= 9600
        self.timeout= 0.1
        self.counter = 0
        
        if self.live_comm : # open serial port
            self.ser = serial.Serial(
                    port=self.port,
                    baudrate=self.baudrate,
                    timeout=self.timeout
                    ) 
            
            self.ser.isOpen()
            self.send("")
            
            # ensure scope position is updated
            self.sidereal_time()
            self.set_latitude(self.lat)
            
        # TODO:
        # 2) implement threading, close/fail gracefully
        # 3) build reader thread for catching responses

        print "ARCHIMAGE __INIT__: INITIALIZED COM INTERFACE!"
    
    '''
    -----------------------------------------------------------------
    /////////////////  High-Level Mount Operations  \\\\\\\\\\\\\\\\\
    -----------------------------------------------------------------
    '''

    def connect(self):
        '''
        Establish serial comms
        '''
        print "connected!"
        return "connected!"

    def disconnect(self):
        '''
        Break serial comms
        '''
        print "disconnected!"
        return "disconnected!"

    def get_all(self):
        '''
        Returns all "get-able" parameters
        '''
        data = {
                'ra': self.get_pointing_ra(),
                'ha': self.get_pointing_ha(),
                'dec': self.get_pointing_dec(),
                'az': self.get_pointing_az(),
                'alt': self.get_pointing_alt(),
                'track_ha': self.get_ha_rate(),
                'track_dec': self.get_dec_rate()
                }

        print json.dumps(data,indent=4)

        return data
        
    def sidereal_time(self):
        '''
        Returns the current sidereal time, and set's the scopes sidereal time
        '''
        sidereal_time = compute_sidereal_time(self.lon)
        return self.set_sidereal(sidereal_time)
        
    def point_radec(self,ra_deg,dec_deg):
        '''
            Set's object position and executes a goto
        '''
        self.set_object_ra(ra_deg)
        self.set_object_dec(dec_deg)
        self.goto()

    def point_altaz(self,alt_deg,az_deg):
        '''
            Executes an Az/el goto
        '''
        self.altaz_goto(alt_deg,az_deg)

    def init_align(self):
        '''
            Initializes mount pointing when pointed south at zero degrees elevation
        '''

        lat = self.get_latitude()
        dec = -1.0*(90.0-lat)
        self.set_pointing_dec(dec)
        self.set_pointing_ha(0.0)
        
        return "aligned south, at zero degrees elevation!"

    def clear_track_rates(self):
        self.set_ha_rate(0)
        self.set_dec_rate(0)
        return "Set HA and DEC rates to zero deg/sec!"

    
    '''
    ---------------------------------------------------------------
    /////////////////  Low-Level Mount Functions  \\\\\\\\\\\\\\\\\
    ---------------------------------------------------------------
        Low-level mount functions are listed here in the following categories:

    1) Mount Flags
    2) Configuration
    3) Alignment and Pointing
    4) Home Stuff
    5) Tracking
    6) Target Specification
    7) Paddle

    '''

    '''
    -------------------------
    1) Mount Flags
    -------------------------

    The mount has several fags (bits) that are use to enable/disable
    features of the mount.  They are outlined below.  I provided a single command
    for each set/clear option.  None of the below commands is checked for compatiability 
    with the remaining flags (for instance, enabling the german equatorial mode while configured as
    an alt/az mount).  Such validation should occur at a higher level.

    Breakdown of flags based on sniffing commmands from MCC Scope:
    _____________________________________________________________
    | Flag |    Name    |                Description            |
    -------------------------------------------------------------
    |   4  | Equatorial | Equatorial vs. Alt/Az                 |
    -------------------------------------------------------------
    |  20  |  Horrizon  | Impose an Elevaiton Limit             |
    -------------------------------------------------------------
    |  23  |    PEC     | Periodic Error Correction             |
    -------------------------------------------------------------
    |  16  |    IECA    | Integrated Error Correction Algorithm |
    -----------------------------------------------------------
    |   0  |  Tracking  | Toggle Tracking                       |
    -------------------------------------------------------------
    |  21  |   German   | German Equatorial                     |
    -------------------------------------------------------------
    '''

    def enable_track(self):
        # turn on tracking
        return self.send("set flag[0]=1")
    
    def disable_track(self):
        # turn off tracking
        return self.send("set flag[0]=0")

    def set_equatorial(self):
        return self.send("set flag[4]=1")

    def set_altaz(self):
        return self.send("set flag[4]=0")

    def enable_german_equatorial(self):
        return self.send("set flag[21]=1")

    def disable_german_equatorial(self):
        return self.send("set flag[21]=0")

    def enable_horrizon_limit(self):
        return self.send("set flag[20]=1")

    def disable_horrizon_limit(self):
        return self.send("set flag[20]=0")

    def enable_pec(self):
        return self.send("set flag[23]=1")

    def disable_pec(self):
        return self.send("set flag[23]=0")

    def enable_ieca(self):
        return self.send("set flag[16]=1")

    def disable_ieca(self):
        return self.send("set flag[16]=0")

    '''
    -------------------------
    2) Configuration
    -------------------------

    These commands configure the mount's latitude and sidereal time (longitude).

    '''
    def set_sidereal(self,decimal_hour_angle):
        '''
            Set the local apparent sidereal time.
            Specify the decimal hour angle.
        '''
        command = "set sidereal=" + "{:.6f}".format(decimal_hour_angle) + "h"
        return self.send(command)
        
    def get_sidereal(self):
        '''
            Return sideral time from the mount as degrees
        '''
        command = "get sidereal"
        resp = self.send(command)
        if self.live_comm:
            return hms2dd(resp['payload'])
        else:
            return random.uniform(0.000,24.000)
    
    def set_latitude(self,decimal_latitude_degrees):
        command = "set latitude=" + "{:.6f}".format(decimal_latitude_degrees) + "d"
        return self.send(command)
        
    def get_latitude(self):
        '''
            Return mount latitude as degrees
        '''
        command = "get latitude"
        resp = self.send(command)
        if self.live_comm:
            return dms2dd(resp['payload'])
        else:
            return random.uniform(-90.000,90.000)

    '''
    -------------------------
    3) Alignment and Pointing
    -------------------------
    Depending on if it's configured as equatorial or alt/az, you can specify the current
    pointing at any time. The commands below set the current pointing in each frame.  They 
    do not check the mounts configuration as equatorial or alt/az.  In equatorial pointing
    is defined in local hour hangle (ha, South is 0 degrees and declination (dec). 
    In alt/az pointing is defned in azimuth (az, North is 0 degrees) and elevation.

    Always handle angles in decimal degrees, however konw that this API translates them to either
    degrees-minutes-seconds or hour angle depending on the parameter being used:

    alt/az pointing is communicated in DMS

    ra pointing is communicated in HMS
    
    dec pointing is communicated in DMS

    '''

    # Equatorial Pointing:
    
    def get_pointing_ra(self):
        command = "get ra"
        resp = self.send(command)
        if self.live_comm:
            return hms2dd(resp['payload'])
        else:
            return random.uniform(0.000,360.000)
        
    def get_pointing_ha(self):
        command = "get ha"
        resp = self.send(command)
        if self.live_comm:
            return hms2dd(resp['payload'])
        else:
            return random.uniform(0.000,360.000)
    
    def get_pointing_dec(self):
        command = "get dec"
        resp = self.send(command)
        if self.live_comm:
            dms2dd(resp['payload'])
        else:
            return random.uniform(-90.000,90.000)

    def set_pointing_ha(self,ha_dd):
        '''
            specifcy pointing in decimal degrees
        '''
        command = "set ha=" + dd2hms(ha_dd)
        return self.send(command)
    
    def set_pointing_dec(self,dec_dd):
        '''
            specifcy pointing in decimal degrees
        '''
        command = "set dec=" + dd2dms(dec_dd)
        return self.send(command)

    # Alt/Az Pointing:
        
    def get_pointing_az(self):
        command = "get az"
        resp = self.send(command)
        if self.live_comm:
            return dms2dd(resp['payload'])
        else:
            return random.uniform(0.000,359.999)
    
    def get_pointing_alt(self):
        command = "get alt"
        resp = self.send(command)
        if self.live_comm:
            return dms2dd(resp['payload'])
        else:
            return random.uniform(-90.000,90.000)

    def set_pointing_az(self,az_dd):
        '''
            specifcy pointing in decimal degrees
        '''
        command = "set az=" + dd2dms(az_dd)
        return self.send(command)
    
    def set_pointing_alt(self,alt_dd):
        '''
            specifcy pointing in decimal degrees
        '''
        command = "set alt=" + dd2dms(alt_dd)
        return self.send(command)

    '''
    -------------------------
    4) Home Stuff
    -------------------------

    These commands configure the mount's home position

    '''
    def find_home(self):
        return self.send("home find")
    def home_setup(self):
        return self.send("home setup")
    def home_park(self):
        return self.send("home park")
    def set_park(self):
        return self.send("home x_parkset")

    '''
    -------------------------
    5) Tracking
    -------------------------

    These commands allow configuration of the tracking rate for each axis.  Axes rates
    are commanded in units of degrees/second.

    '''
    def get_ha_rate(self):
        '''
            return the ha-axis rate in deg/s
        '''
        command = "get trackha"
        resp = self.send(command)
        if self.live_comm:
            return float(resp['payload'])
        else:
            return random.uniform(0.000,4.000)

    def get_dec_rate(self):
        command = "get trackdec"
        resp = self.send(command)
        if self.live_comm:
            return float(resp['payload'])
        else:
            return random.uniform(0.000,4.000)

    def set_ha_rate(self,rate_ds):
        '''
            Specifcy ha tracking rate in deg/s
            set trackha=+0.047314d
        '''
        command = "set trackha=" +  "{:.5f}".format(rate_ds) + "d"
        return self.send(command)
    
    def set_dec_rate(self,rate_ds):
        '''
            Specifcy dec tracking rate in deg/s
            set trackdec=-0.115978d
        '''
        command = "set trackdec=" + "{:.5f}".format(rate_ds) + "d"
        return self.send(command)

    '''
    -------------------------
    6) Target Specification
    -------------------------

    These commands are used to aquire targets in ra/dec or az/el coordinates

    '''
         
    # set/get target parameters
    
    def set_object_ra(self,hour_angle_dd):
        command = "set objectra=" + "{:.5f}".format(hour_angle/15) + "h"
        return self.send(command)
        
    def get_object_ra(self):
        command = "get objectra"
        resp = self.send(command)
        return hms2dd(resp['payload'])
        
    def set_object_dec(self,declination_dd):
        command = "set objectdec=" + "{:.5f}".format(declination) + "d"
        return self.send(command)
        
    def get_object_dec(self):
        command = "get objectdec"
        resp = self.send(command)
        return dms2dd(resp['payload'])
    
    def get_object_az(self):
        command = "get objectaz"
        resp = self.send(command)
        return dms2dd(resp['payload'])
        
    def get_object_alt(self):
        command = "get objectalt"
        resp = self.send(command)
        return dms2dd(resp['payload'])

    def goto(self):
        return self.send("goto")
    
    def altaz_goto(self,alt,az):
        return self.send("goto objectaz="+dd2dms(az)+" objectalt="+dd2dms(alt))   
    '''
    -------------------------
    7) Paddle
    -------------------------

    These commands are built in for jogging the mount.

    '''
    def set_speed(self,slew_speed_ds):
        command = "set speed=" + "{:.5f}".format(slew_speed_ds) + "d"
        return self.send(command)
        
    def move_north(self):
         return self.send("move north")
         
    def stop_north(self):
         return self.send("stop north")
         
    def move_east(self):
         return self.send("move east")
         
    def stop_east(self):
         return self.send("stop east")
         
    def move_south(self):
         return self.send("move south")
         
    def stop_south(self):
         return self.send("stop south")
         
    def move_west(self):
         return self.send("move west")
         
    def stop_west(self):
         return self.send("stop west")
         
    def stop(self):
         return self.send("stop")

    '''
    -----------------------------------------------------------
    /////////////////  Serial Comm functions  \\\\\\\\\\\\\\\\\
    -----------------------------------------------------------
    '''
            
    def send(self,payload,verbose=True):
        '''
        This accepts command strings as readable text (ascii).
        It creates a packet by adding header, footer, counter, check bytes.
        It then fires it off to the serial port (configured by the archimage class)
        
        TODO:
        1) Add a blocking read to catch the reponse after each message.
        '''
        if verbose:
            print "[TX_SEND]", payload

        if self.live_comm :
            # Serial Purge:
            # I noticed this while sniffing the CSAT serial comms.  Seems to help.
            while self.ser.out_waiting > 0 :
                self.ser.reset_input_buffer()
            self.ser.write(binascii.unhexlify("08000000"))
            # format the packet
            dbug(self.debug_flag,'-----------------------')
            dbug(self.debug_flag,'Payload: ' + payload)
            hex_payload = prepare(payload,self.counter)
            dbug(self.debug_flag,'Hex: ' + hex_payload)
            bytes = binascii.unhexlify(hex_payload)
            # send the packet
            self.ser.write(bytes)
            # up the counter
            self.counter = countup(self.counter)
            dbug(self.debug_flag,'Counter: ' + str(self.counter))
            response = unpack(self.ser.read(1000))
            #check for "ok" or "err"
            rx_payload = ''
            if response[0:2] == 'ok':
                status = "success"
                error_flag = False
                parsed = re.findall(r'ok (.*?)\r',response,re.DOTALL)
                if parsed:
                    rx_payload = rx_payload + parsed[0]
            elif response[0:5] == 'error':
                status = "error"
                error_flag = True
                parsed = re.findall(r'(.*?)\r',response,re.DOTALL)
                if parsed:
                    rx_payload = rx_payload + parsed[0]
            else:
                status = "unknown"
                error_flag = False
                rx_payload = rx_payload + response

        else: # not live
            rx_payload = 'DEAD REPLY...'
            error_flag = False
            status = "success"
            response = "no bytes fool"

        if verbose:
            print '[RX_STATUS]', status
            print '[RX_RESPONSE]', rx_payload

        reply = {}
        reply['error'] = error_flag
        reply['status'] = status
        reply['payload'] = str(rx_payload.decode("utf8"))
        reply['raw'] = response
        return reply
           

'''
-----------------------------------------------
/////////////////  Utilities  \\\\\\\\\\\\\\\\\
-----------------------------------------------
'''

def dd2dms(degrees):
    '''
        given angle as degree (some kind of number...) return signed dms string
    '''
    a = ephem.degrees(str(degrees))
    dms = str(a).replace(':','d',1)
    dms = dms.replace(':','m')+'s'
    if a >= 0:
        dms = "+" + dms
    return dms

def dd2hms(degrees):
    '''
        given angle as degree (some kind of number...) return signed hms string
    '''
    a = ephem.hours(ephem.degrees(str(degrees)))
    hms = str(a).replace(':','h',1)
    hms = hms.replace(':','m')+'s'
    if a >= 0:
        hms = "+" + hms
    return hms

def dms2dd(dms):
    '''
        given signed dms string return float degrees
    '''
    dms = dms.strip()
    dms = dms.replace('d',':')
    dms = dms.replace('m',':')
    dms = dms.replace('s','')
    a = ephem.degrees(dms)
    dd = a*180/ephem.pi
    return dd

def hms2dd(hms):
    '''
        given signed hms string return float degrees
    '''
    hms = hms.strip()
    hms = hms.replace('h',':')
    hms = hms.replace('m',':')
    hms = hms.replace('s','')
    a = ephem.hours(hms)
    dd = a*180/ephem.pi
    return dd

def dbug(flag,text):
    if flag:
        print "[DBUG]: ",text
        
def countup(counter):
    '''
        The following increments in the counter in DECIMAL.
        I based this counting scheme on what I observed during comms capture
    '''
    if counter == 0:
        counter = 17
        return counter
    else:
        counter = counter + 16;
        if counter > 255:
            counter = 1
        return counter
        
def prepare(payload,Counter):
    Payload = payload.encode('hex')
    Header = '1002'
    SOH = '01'
    Counter = '{:02x}'.format(Counter)
    Filler = checkbytes(fletcher_16(SOH+Counter+Payload))
    Footer = '1003'
    return Header+SOH+Counter+Payload+Filler+Footer
    
def unpack(hex_response):
    # Get rid of the first 4 bytes:
    # hex_payload = hex_response[4:-4]
    hex_payload = bytearray(hex_response)
    return hex_payload[4:-4]
 
def fletcher_16(hex_string, modulus=255):
	'''
	Calculate the Fletcher-16 checksum of *data*, default modulus 255.

	Returns an upper case, zero-filled hex string with no prefix such as
	``0A1B``.

	>>> fletcher_16("hello,world")
	'6C62'
	>>> fletcher_16("hello,world", 256)
	'6848'
	'''

	numbers = bytearray.fromhex(hex_string)

	a = b = 0
	for number in numbers:
		a += number
		b += a
	a %= modulus
	b %= modulus
	fletch = hex((a << 8) | b)[2:].upper().zfill(4)
	return fletch

def checkbytes(f16, modulus=255):
	'''
	 Accepts an upper case, zero-filled hex string with no prefix.
	 Assumes a 2-byte fletcher checksum is given...
	'''
	c0 = int(f16[:2],16)
	c1 = int(f16[-2:],16)
	cb0 = modulus - ((c0 + c1)%modulus)
	cb1 = modulus - ((c0 + cb0)%modulus)
	hex_string = '{0:02x}'.format(cb0)+'{0:02x}'.format(cb1)
	return hex_string.upper()
    
def compute_sidereal_time(lon,lat=0,alt=0,t=datetime.utcnow()):
    '''
        Return local apparent sidereal time in decimal hours:
        
        Inputs [required]
        lon - (float) longitude in decimal degrees
        
        Inputs [optional]
        lat - (float, default=0) latitude in decimal degrees
        alt - (float, default=0) altitude in meteres
        time - (float, default=now) datetime object
        
        Output
        sidereal_time - (float) local apparent sidereal time in decimal hours
    '''
    ovr = ephem.Observer()
    ovr.lon = lon * ephem.degree
    ovr.lat = lat * ephem.degree
    ovr.elevation = alt
    ovr.date = t
    st = ovr.sidereal_time()
    st_hours = (st/ephem.degree)/15 # convert time in radians to decimal hours
    return st_hours

'''
----------------------------------------------------------
/////////////////  Command Line Console  \\\\\\\\\\\\\\\\\
----------------------------------------------------------
'''
def Console(live=True):
    '''
        Provides console interface to talk to the mount.  Takes care of command formatting.
        Type "exit" to end.
    '''
    print "-------------------------------------------------------------"
    print "           Archimage Telescope Command Console"
    print "-------------------------------------------------------------"
    print " + Type commands and press <enter> to send.\n + Responses will be displayed after each command.\n + Type 'exit' and press <enter> to quit."
    print "-------------------------------------------------------------"

    # instantiate mount, set live/testing
    mount = archimage(live=live)

    while True:
        print "Enter command:"
        p = raw_input(">> ")
        if p == 'exit':
            print "-------------------------------------------------------------"
            print "                          Bye."
            print "-------------------------------------------------------------"
            exit()
        else:
            mount.send(p)
    
if __name__ == "__main__":
    Console(live=True)
    